worksheet.rb

```diff
@@ -31,4 +77,100 @@
 # - the total amount of money each driver has made
 # - the average rating for each driver
 # - Which driver made the most money?
-# - Which driver has the highest average rating?
\ No newline at end of file
+# - Which driver has the highest average rating?
+
+# Define data structures
+average_ratings_array = []
+top_earning_drivers = []
+highest_rated_drivers = []
+
+# Define methods
+# Method to store the highest-achieving drivers in given categories
+def find_the_highest(array_to_compare, array_of_highest, hash)
+  
+  array_to_compare.each_with_index do |item, index|
+    if item == array_to_compare.max
+      array_of_highest << hash.keys[index]
+    end
+  end
+  
+end
+
+# Method to print the highest-achieving drivers in given categories
+def print_highest_results (array_of_highest, category_string, plural_string, singular_string)
+  
+  if array_of_highest.length > 1
+    print "#{category_string}#{plural_string}"
+    (array_of_highest.length - 1).times do |i|
+      print "#{array_of_highest[i]}, "
+    end
+    puts "and #{array_of_highest.last}."
+  else
+    puts "#{category_string}#{singular_string}#{array_of_highest.first}."
+  end
+  
+end
+
+# - The number of rides each driver has given
+data_hash.each do |driver, info|
+  puts "#{driver} gave #{info[:costs].length} rides."
+end
+
+# - The total amount of money each driver has made
+total_money_array = []
+
+data_hash.each do |driver, info|
+  total_money = info[:costs].sum
+  puts "#{driver} made $#{total_money} total."
+  total_money_array << total_money
+end
+
+# - The average rating for each driver
+data_hash.each do |driver, info|
+  average_rating = (info[:ratings].sum.to_f) / (info[:ratings].length.to_f)
+  puts "#{driver} had an average rating of #{average_rating}."
+  average_ratings_array << average_rating
+end
+
+# - Which driver made the most money?
+find_the_highest(total_money_array, top_earning_drivers, data_hash)
+
+print_highest_results(top_earning_drivers, "The top earning", " drivers were ", " driver was ")
+
+# - Which driver has the highest average rating?
+find_the_highest(average_ratings_array, highest_rated_drivers, data_hash)
+
+print_highest_results(highest_rated_drivers, "The highest rated", " drivers were ", " driver was ")
+
+# Optional: For each driver, on which day did they make the most money?
+data_hash.each do |driver, info|
+  
+  highest_date = []
+  third_money = 0
+  fourth_money = 0
+  fifth_money = 0
+  
+  dates_array = ["3rd Feb 2016", "4th Feb 2016", "5th Feb 2016"]
+  
+  info[:dates].each_with_index do |item, index|
+    if item == dates_array[0]
+      third_money += info[:costs][index]
+    elsif item == dates_array[1]
+      fourth_money += info[:costs][index]
+    else
+      fifth_money += info[:costs][index]
+    end
```

Comment:
Nice! This optional problem was tricky! Hm, in an another refactoring, I would love to see this refactored so that `dates_array` was not hardcoded dates, but could be more generic.